File: js/test/catalog.test.js
// Catalog tests. Uses tests/fixtures/catalog-test.yaml so we don't depend on
// the live catalog/projects.yaml.

import test from "node:test";
import assert from "node:assert/strict";
import { mkdtemp, copyFile, readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname, resolve, join } from "node:path";
import { tmpdir } from "node:os";

import { Catalog, loadCatalog } from "../src/catalog.js";

const here = dirname(fileURLToPath(import.meta.url));
const FIXTURE = resolve(here, "../../tests/fixtures/catalog-test.yaml");

test("loadCatalog reads the test fixture", async () => {
  const cat = await loadCatalog(FIXTURE);
  assert.equal(cat.projects.length, 3);
  const ids = cat.projects.map((p) => p.id).sort();
  assert.deepEqual(ids, ["clock", "dreamspace", "realmwatch"]);
});

test("Catalog.resolve finds by id", async () => {
  const cat = await loadCatalog(FIXTURE);
  const proj = cat.resolve("clock");
  assert.ok(proj);
  assert.equal(proj.current_name, "clock.realm.watch");
});

test("Catalog.resolve finds by current_name", async () => {
  const cat = await loadCatalog(FIXTURE);
  const proj = cat.resolve("clock.realm.watch");
  assert.ok(proj);
  assert.equal(proj.id, "clock");
});

test("Catalog.resolve finds by prior name", async () => {
  const cat = await loadCatalog(FIXTURE);
  const proj = cat.resolve("dreamscape");
  assert.ok(proj);
  assert.equal(proj.id, "dreamspace");
  assert.equal(proj.current_name, "dreamscape.realm.watch");
});

test("Catalog.resolve returns null for unknown name", async () => {
  const cat = await loadCatalog(FIXTURE);
  assert.equal(cat.resolve("nonsense"), null);
});

test("Catalog.byRealm filters by realm", async () => {
  const cat = await loadCatalog(FIXTURE);
  assert.equal(cat.byRealm("signal").length, 1);
  assert.equal(cat.byRealm("oracle").length, 1);
  assert.equal(cat.byRealm("none").length, 0);
});

test("Catalog.byStatus filters by status", async () => {
  const cat = await loadCatalog(FIXTURE);
  assert.equal(cat.byStatus("active").length, 2);
  assert.equal(cat.byStatus("local-only").length, 1);
});

test("Catalog.claim with renamesOf appends prior_name and updates current_name", async () => {
  const cat = await loadCatalog(FIXTURE);
  const updated = cat.claim("oracle.realm.watch", {
    renamesOf: "realmwatch",
    reason: "moved into family",
    retired: "2026-05-07",
  });
  assert.equal(updated.current_name, "oracle.realm.watch");
  assert.equal(updated.prior_names.length, 1);
  assert.equal(updated.prior_names[0].name, "realmwatch");
  assert.equal(updated.prior_names[0].reason, "moved into family");
  // resolve via either old or new should still find it
  assert.ok(cat.resolve("realmwatch"));
  assert.ok(cat.resolve("oracle.realm.watch"));
});

test("Catalog.claim adds a fresh entry when no rename", async () => {
  const cat = await loadCatalog(FIXTURE);
  const fresh = cat.claim("brand-new-thing", { reason: "" });
  assert.equal(fresh.current_name, "brand-new-thing");
  assert.equal(fresh.id, "brand-new-thing");
  assert.deepEqual(fresh.prior_names, []);
});

test("Catalog.claim refuses duplicate fresh claim", async () => {
  const cat = await loadCatalog(FIXTURE);
  assert.throws(() => cat.claim("clock"));
});

test("Catalog.toYaml round-trip is parseable and preserves count", async () => {
  const cat = await loadCatalog(FIXTURE);
  const yamlText = cat.toYaml();
  const dir = await mkdtemp(join(tmpdir(), "lexicon-cat-"));
  const path = join(dir, "out.yaml");
  await import("node:fs/promises").then((fs) =>
    fs.writeFile(path, yamlText, "utf8"),
  );
  const reloaded = await loadCatalog(path);
  assert.equal(reloaded.projects.length, cat.projects.length);
  const ids = reloaded.projects.map((p) => p.id).sort();
  assert.deepEqual(ids, ["clock", "dreamspace", "realmwatch"]);
});

test("Catalog.save writes to disk", async () => {
  const dir = await mkdtemp(join(tmpdir(), "lexicon-cat-save-"));
  const path = join(dir, "projects.yaml");
  await copyFile(FIXTURE, path);
  const cat = await loadCatalog(path);
  cat.claim("brand-new", { reason: "" });
  await cat.save();
  const text = await readFile(path, "utf8");
  assert.match(text, /brand-new/);
});

File: go/catalog_test.go
package lexicon

import (
	"path/filepath"
	"testing"
)

func loadTestCatalog(t *testing.T) *Catalog {
	t.Helper()
	c, err := LoadCatalog(filepath.Join(testdataDir(t), "catalog-test.yaml"))
	if err != nil {
		t.Fatalf("LoadCatalog: %v", err)
	}
	return c
}

func TestLoadCatalog_BasicShape(t *testing.T) {
	c := loadTestCatalog(t)
	if got, want := len(c.Projects), 3; got != want {
		t.Errorf("len(Projects) = %d, want %d", got, want)
	}
	clock := c.Projects[0]
	if clock.ID != "clock" || clock.CurrentName != "clock.realm.watch" {
		t.Errorf("clock entry malformed: %+v", clock)
	}
}

func TestLoadCatalog_PriorNamesParsed(t *testing.T) {
	c := loadTestCatalog(t)
	dream := c.Projects[1]
	if got, want := len(dream.PriorNames), 2; got != want {
		t.Errorf("len(PriorNames) = %d, want %d", got, want)
	}
	if dream.PriorNames[0].Name != "dreamspace" {
		t.Errorf("first prior name = %q, want dreamspace", dream.PriorNames[0].Name)
	}
}

func TestLoadCatalog_NullableFields(t *testing.T) {
	c := loadTestCatalog(t)
	rw := c.Projects[2]
	if rw.Domain != "" {
		t.Errorf("realmwatch.domain should be empty, got %q", rw.Domain)
	}
	if rw.Repo != "" {
		t.Errorf("realmwatch.repo should be empty, got %q", rw.Repo)
	}
}

func TestLoadCatalog_AudienceParsed(t *testing.T) {
	c := loadTestCatalog(t)
	if got := c.Projects[0].Audience; got != "realm" {
		t.Errorf("clock.audience = %q, want %q", got, "realm")
	}
	// dreamspace and realmwatch fixtures don't set audience → empty.
	if got := c.Projects[1].Audience; got != "" {
		t.Errorf("dreamspace.audience = %q, want empty (not set in fixture)", got)
	}
}

func TestLoadCatalog_Missing(t *testing.T) {
	_, err := LoadCatalog("/no/such.yaml")
	if err == nil {
		t.Fatal("expected error")
	}
}
